src/diff_xdiff.c

```diff
@@ -115,13 +116,17 @@ static int git_xdiff_cb(void *priv, mmbuffer_t *bufs, int len)
 	const git_diff_delta *delta = patch->base.delta;
 	git_patch_generated_output *output = &info->xo->output;
 	git_diff_line line;
+	int buffer_len;
 
 	if (len == 1) {
 		output->error = git_xdiff_parse_hunk(&info->hunk, bufs[0].ptr);
 		if (output->error < 0)
 			return output->error;
 
-		info->hunk.header_len = bufs[0].size;
+		/* Sanitize the hunk header in case there is invalid Unicode */
+		buffer_len = git__utf8_valid_buf_length((const uint8_t *) bufs[0].ptr, bufs[0].size);
+		info->hunk.header_len = buffer_len;
```

Comment:
So what you're doing here is cutting off at the point where there is the first invalid UTF8 codepoint. I think this solution here is actually not sufficient yet. In case `info->hunk.header_len >= sizeof(info->hunk.header)`, you're actually stripping the last byte again. In case the last byte belongs to a multi-byte  codepoint again, it will become invalid. So you actually have to sanitize after checking the header length.

But stepping back a bit: is it actually guaranteed that the header is in UTF8? Git is agnostic to the encoding in which files are being stored, so I could envision the diff machinery processing, say, UTF16-encoded data.

Reply:
Git doesn't do well with UTF16.  UTF16 files will be treated as binary (since there's a NUL) and will probably not have a header at all.  I'm curious what would happen with non-UTF8 multibyte charsets, but I think it's good enough to emulate Git's behavior, which also seems to punt on those.

Reply:
Yeah, I know about them being treated as binary by default. It's possible to override that via gitattributes, though. But as you say, it's probably okay to just punt on that, then. It's only non-critical metainformation, after all.